pkg/bundle/sync.go

```diff
@@ -165,24 +174,61 @@ func encodeJKS(trustBundle string) ([]byte, error) {
 
 		c, err := x509.ParseCertificate(p.Bytes)
 		if err != nil {
-			return []byte{}, err
+			return nil, fmt.Errorf("got invalid cert when trying to encode JKS: %w", err)
 		}
 
-		ks.SetTrustedCertificateEntry(c.Issuer.String(), jks.TrustedCertificateEntry{
+		// Note on CreationTime:
+		// Debian's JKS trust store sets the creation time to match the time that certs are added to the
+		// trust store (i.e., it's effectively time.Now() at the instant the file is generated).
+		// Using that method would make our JKS files in trust-manager non-deterministic, leaving us with
+		// two options if we want to maintain determinism:
+		// - Using something from the cert being added (e.g. NotBefore / NotAfter)
+		// - Using a fixed time (i.e. unix epoch)
+		// We use NotBefore here, arbitrarily.
+
+		alias := jksAlias(c)
+
+		err = ks.SetTrustedCertificateEntry(alias, jks.TrustedCertificateEntry{
 			CreationTime: c.NotBefore,
 			Certificate: jks.Certificate{
 				Type:    "X509",
 				Content: p.Bytes,
 			},
 		})
+
+		if err != nil {
+			// this error should never happen if we set jks.Certificate correctly
+			return nil, fmt.Errorf("failed to add cert with alias %q to trust store: %w", alias, err)
+		}
 	}
 
 	buf := &bytes.Buffer{}
-	ks.Store(buf, []byte(defaultJKSPassword))
+
+	err := ks.Store(buf, password)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create JKS file: %w", err)
+	}
 
 	return buf.Bytes(), nil
 }
 
+// jksAlias creates a JKS-safe alias for the given certificate, such that
+// any two certificates will have a different aliases unless they're
+// identical in every way.
+// This unique alias fixes an issue where we used the Issuer field as an
+// alias, leading to different certs being treated as identical.
+func jksAlias(cert *x509.Certificate) string {
+	certHashBytes := sha256.Sum256(cert.Raw)
+	certHash := hex.EncodeToString(certHashBytes[:])
+
+	// Since certHash is the part which actually distinguishes between two
+	// certificates, put it first so that it won't be truncated if a cert
+	// with a really long subject is added. Not sure what the upper limit
+	// for length actually is, but it shouldn't matter here.
+
+	return certHash[:8] + "|" + cert.Subject.String()
+}
+
```

Comment:
note: this is the real core of the bug fix, and this is what fixes the reported issue.

For comparison, the Debian JKS file uses the filename that a trusted cert was taken from as the alias, which is obviously a unique measure. We don't have filenames, but this approach of using a hash seems to work well.

To check the JKS truststore on a debian based machine with Java installed:

```console
$ keytool -list -keystore /etc/ssl/certs/java/cacerts
<snip>
...
debian:verisign_universal_root_certification_authority.pem, 2 May 2023, trustedCertEntry,
Certificate fingerprint (SHA-256): 23:99:56:11:27:A5:71:25:DE:8C:EF:EA:61:0D:DF:2F:A0:78:B5:C8:06:7F:4E:82:82:90:BF:B8:60:E8:4B:3C
debian:xramp_global_ca_root.pem, 2 May 2023, trustedCertEntry,
Certificate fingerprint (SHA-256): CE:CD:DC:90:50:99:D8:DA:DF:C5:B1:D2:09:B7:37:CB:E2:C1:8C:FB:2C:10:C0:FF:0B:CF:0D:32:86:FC:1A:A2
```